src/main/java/gui/viewmodel/TodoListPageViewModel.java

```diff
@@ -3,30 +3,74 @@
 import datagateway.task.ObservableTaskRepository;
 import datagateway.task.TaskReader;
 import javafx.collections.FXCollections;
-import javafx.collections.ObservableMap;
+import javafx.collections.ObservableList;
 
+import java.time.LocalDateTime;
 import java.time.format.DateTimeFormatter;
 import java.time.format.FormatStyle;
-import java.util.List;
+import java.util.*;
 
+/***
+ * taskReaderList: a sorted list based on the deadline of tasks.
+ *                 tasks without deadline will be put to the end of the list
+ * taskInfoList: a list of maps of task info corresponding to taskReaderList
+ */
 public class TodoListPageViewModel extends ViewModel{
 
     private final ObservableTaskRepository repository;
 
-    private final ObservableMap<String, String> taskInfoMap;
+    private final ObservableList<TaskReader> taskReaderList;
+    private final ObservableList<Map<String, String>> taskInfoList;
 
     public TodoListPageViewModel(ObservableTaskRepository repository) {
         this.repository = repository;
         repository.addCreationObserver(this::handleCreation);
         repository.addUpdateObserver(this::handleUpdate);
 
+        // initialize sorted taskReaderList
         List<TaskReader> taskList = repository.getAllTasks().get(0L);
-        this.taskInfoMap = generateTaskNameDeadlineMap(taskList);
+        this.taskReaderList = FXCollections.observableArrayList();
+        for (TaskReader taskReader : taskList) {
+            this.insertTaskReader(taskReader);
+        }
+        // initialize taskInfoList accordingly
+        this.taskInfoList = FXCollections.observableArrayList();
+        this.updateTaskInfoList();
     }
 
-    private ObservableMap<String, String> generateTaskNameDeadlineMap(List<TaskReader> taskList) {
-        ObservableMap<String, String> nameDeadlineMap = FXCollections.observableHashMap();
-        for (TaskReader taskReader : taskList) {
+    /***
+     * helper function to sort the taskReaderList based on the deadline
+     * @param taskReader to inserted to taskReadList
+     */
+    private void insertTaskReader(TaskReader taskReader) {
```

Comment:
For the sorting part, we could use java's default sorting with lambda expression

Reply:
Right, but since there might be non-deadline tasks, I did it manually since I'm not sure whether lambda can deal with this.

Reply:
ok. If lambda does not work we can leave it like this